src/com/github/Kraken3/AFKPGC/LagScanner.java

```diff
@@ -0,0 +1,200 @@
+package com.github.Kraken3.AFKPGC;
+
+import org.bukkit.Location;
+import org.bukkit.Chunk;
+import org.bukkit.World;
+import org.bukkit.block.BlockState;
+import org.bukkit.Material;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.EntityType;
+import java.util.Map;
+import java.util.HashMap;
+
+/**
+ * Should be runnable with callback. First iteration will be straight mainthread code, b/c testing.
+ * 
+ * TODO: Make a runnable, redesign BotDetector to play well with deferred answers.
+ * 
+ * @author ProgrammerDan
+ */
+public class LagScanner {
+
+	private static Map<String, Map<Long, LagScanner.Result>> cache = 
+			new HashMap<String, Map<Long, LagScanner.Result>>();
+
+	public static long cacheTimeout;
+	public static long lagSourceThreshold;
+
+	private Location center;
+	private Integer radius;
+	private ScanCallback<Boolean> callback;
+
+	private Boolean isLagSource;
+	private Long lagCompute;
+
+	public LagScanner(Location center, Integer radius, ScanCallback<Boolean> callback) {
+		this.center = center;
+		this.radius = radius;
+		this.callback = callback;
+		this.isLagSource = null;
+		this.lagCompute = null;
+	}
+
+	public static LagScanner instance(Location center, Integer radius, ScanCallback<Boolean> callback ){
+		return new LagScanner(center, radius, callback);
+	}
+
+	public void run() {
+		boolean lagSource = false;
+		Chunk originChunk = center.getChunk();
+		World chunkWorld = originChunk.getWorld();
+		int oX = originChunk.getX();
+		int oZ = originChunk.getZ();
+		long now = System.currentTimeMillis(); // all tests against same millis ... for now.
+		long lagSum = 0L;
+		int chunksTested = 0;
+
+		// radius in minecraft is square.
+		for (int r = 0; r <= radius && !lagSource; r++) {
+			for (int x = oX - r; x <= oX + r && !lagSource; x++) {
+				for (int z = oZ - r; z <= oZ + r; z++) {
+					if (x > oX - r && x < oX + r &&
+						z > oZ - r && z < oZ + r) {
+						continue; // don't retread inner chunks of prior radius

```

Comment:
Wait yeah, I feel stupid. I've never seen something like this before, quick question completly unrelated to this and to minecraft in general: Is there any advantage to doing it this way from an algorithm view opposed to the classic row by row?